### ISSUE TITLE
Rewrite AI

### ISSUE THREAD

**Issue by stefanhendriks** [stefanhendriks/Dune-II---The-Maker#171]
This issue was originally created from "I noticed the AI is not attacking anymore?" -> AI is too passive.

Instead of trying to figure out why it does not work, a better way would be to sit down and write how I want it to work. Since the code of the AI is very old, refactored a bit, but still hinges on old patterns. 

### TODO

- [x] Make sure AI for Fremen Super Weapon works
- [x] Sand-worm AI
- [x] Make AI brain pluggable, not full with if's based on "player id" and such
- [x] Dune 2 Scenario AI
    - [x] Different skill levels? (deferred to issue #189)
    - [x] Use upgrades/upgrade where needed? (no)
    - [x] Re-build destroyed structures? (or keep it delegated to #139 ? - decide)
    - [x] #208
    - [x] AI should be able to do anything, even though it is not technically allowed to do so. (ie, meaning, it may cheat, defeating #135 in a way, but for campaign missions this is unavoidable).
        - [x] Fix out of power by cheating (ie fix #147 in Campaign mode)
        - [x] rebuild all structures required (ie Mission 9)
    - [x] Whenever the AI builds a structure, it should have slabs beneath it (just like in Dune 2)
    - [x] Build army and attack after x time
    - [x] Only be active after having discovered human player
    - [x] Build army and attack (only once we discovered human player)
    - [x] Use different kind of 'groups' to attack the player, depending on the mission.
    - [x] Let AI improve their economy (2 harvesters per refinery, 1 carry-all per refinery if possible).
    - [x] #207
    - [x] When AI had multiple missions, re-add finished missions. (don't wait for all missions to be over, but allow multiple missions to be active.
    - [x] AI should be able to use deathhand
    - [x] AI should be able to use saboteur
    - [x] AI should be able to use fremen
        - [x] The Fremen should be friendly to the AI
- [x] D2TM Skirmish AI
    - [x] #205 (deferred)
    - [x] Build base
    - [x] Build army
    - [x] Use upgrades
    - [x] Build defenses? (Turrets/Rocket turrets?)
    - [x] Let AI build up proper economy
    - [x] #206 (deferred)
    - [x] Use Super weapons
    - [x] When in end-game, do kamikaze with all remaining units. (mainly so that the player does not have to scout/search for remaining units...)

- [x] run through Deleaker and see if any memory issues are present, if so fix them

## Playtesting bugs found
- [x] Enemy detected triggers music, even when human does not detect enemy?
- [x] #212 (defer, playtesting bug)
- [x] #209 (defer, playtesting bug)
- [x] #210 (defer, playtesting bug)
- [x] #211 (defer, playtesting bug)

**Comment by stefanhendriks** [comment]
I feel like this would be even better to rewrite parts of the AI instead. It is a PITA to find anything in this old code.

**Comment by stefanhendriks** [comment]
Did some playtesting, the AI is now very aggressive in Campaign mode. So it probably needs tweaking before actual release.
